modules/mod-deep-learning/DeepLearningAnalyzer.cpp

```diff
@@ -0,0 +1,232 @@
+
+      // if we're not doing a multichannel forward pass, downmix
+      if (!mModel->GetCard()->multichannel())
+         input = input.sum(0, true, torch::kFloat);
+
+      // forward pass!
+      torch::jit::IValue output = ForwardPassInThread(input);
+
+      // split forward pass output into output tensor and timestamps
+      auto [modelOutput, timestamps] = mModel->ToTimestamps(output);
+
+      // write the block's label to the label track
+      double blockStart = leader->LongSamplesToTime(samplePos);
+      sampleCount blockEndSamples = samplePos + (sampleCount)blockSize;
+      double blockEnd = leader->LongSamplesToTime(blockEndSamples);
+
+      TensorToLabelTrack(modelOutput, labelTrack, blockStart, blockEnd, timestamps);
+
+      // Update the Progress meter
+      double tPos = leader->LongSamplesToTime(samplePos);
+      if (TrackProgress(mCurrentTrackNum, (tPos - tStart) / (tEnd - tStart)))
+      return false;
+   }
+   labelTrack->Commit();
+   return true;
+}
+
+std::vector<Stamp> Coalesce(std::vector<Stamp> &timestamps)
+{
+   if (timestamps.size() == 0)
+      return timestamps;
+
+   // sort the stamps by start time
+   std::function< bool(Stamp &a, Stamp &b)> compareByStartTime (
+      [](Stamp &a, Stamp &b)
+      {
+         return a.first < b.first;
+      }
+   );
+
+   std::sort(timestamps.begin(), timestamps.end(), compareByStartTime);
+   std::vector<Stamp> coalescedStamps; 
+
+   Stamp prev = timestamps[0];
+   for (const auto &stamp : timestamps)
+   {
+      if (stamp.first <= prev.second)
+      {
+         if (stamp.second > prev.second)
+         {
+            prev.second = stamp.second;
+         }
+      }
+      else
+      {
+         coalescedStamps.emplace_back(Stamp(prev));
+         prev = Stamp(stamp);
+      }
+   }
+
+   // check if last is not equal to end of last stamp
+   // append the last stamp
+   if (!coalescedStamps.empty())
+   {
+      if (coalescedStamps.back().second != prev.second)
+      {
+         coalescedStamps.emplace_back(prev);
+      }
+   }
+   else if (coalescedStamps.empty() && !timestamps.empty())
+   {
+      coalescedStamps.emplace_back(prev);
+   }
+
+   return coalescedStamps;
+}
+
+// TODO: coalesce labels
+void DeepLearningAnalyzer::TensorToLabelTrack
+(torch::Tensor output, std::shared_ptr<AddedAnalysisTrack> labelTrack,
+   double tStart, double tEnd, torch::Tensor timestamps) 
+{
+   // add the start offset to the timestamps
+   timestamps += tStart;
+
+   // check that the output is one dimensional
+   wxASSERT(output.dim() == 1);
+
+   // create a map of labels to timestamps
+   std::map<wxString, std::vector<Stamp> > timestampsMap;
+
+   for (int i = 0 ; i < output.size(0); i++)
+   {
+      torch::Tensor classIndex = output[i];
+      wxString label = mClasses[classIndex.item().to<int>()];
+      
+      timestampsMap[label].emplace_back(
+         Stamp(timestamps[i][0].item().to<double>(),
+              timestamps[i][1].item().to<double>())
```

Comment:
`Stamp(...)` is unnecessary; `emplace_back` itself is a variadic function passing arguments to the element type constructor.